docs: Add theory for sections and materials (#197)
fib-international/structuralcodes: docs/theory/section_calculator.md

3. **Integrate stresses for all computed ultimate strain profiles**: for all the strain profiles determined in step 2, the stresses are integrated computing the pair $(N, M_y^*)$ that represent a point in the ultimate domain profile.
This is represented by the following image that shows for a rectangular RC section the ultimate strain profiles computed and the corresponding MN interaction domain. The different colors correspond to different domains. The same is visible in the animation [here](theory-anim-domain_nm)

(theory-fig-domain_nm)=
:::{figure} Figure_domain_mn.png

Ultimate strain profiles for uniaxial bending and corresponding interaction domain.
:::

(theory-anim-domain_nm)=
:::{figure} Figure_domain_mn.gif

Animation of ultimate strain profiles for uniaxial bending and corresponding interaction domain.
:::

4. **Rotate back to original $yz$ coordinate system**: the pair of values $(N, M_y^*)$ is rotated back obtaining the triplet $(N, M_y, M_z)$

(theory-nmm-domain)=
## Compute MNN interaction domain

> with this algorithm, *structuralcodes* computes the full three dimensional $N$-$M_y$-$M_z$ interaction domain.

Reviewer: Fix upper case letter.

```suggestion
With this algorithm, *structuralcodes* computes the full three dimensional $N$-$M_y$-$M_z$ interaction domain.
```

For this reason the computation of MM interaction domain for a given value of axial load $N$ does not work in the same way. The side effect of this algorithm is that it is significantly slower since each point of the failure domain is determined with an iterative process.
::: 

:::::{hint}
Using this method it is possible to loop from minimum to maximum axial load that the section is able to carry and construct a nice structured version of the failure domain. Of course this operation is pretty slow compared to the direct computation of [NMM domain](theory-nmm-domain).

::::{dropdown-syntax}
:::{literalinclude} ../_example_code/usage_compute_3d_domain_from_mm.py
:lines: 3-
:caption: Create many slices of MM interaction domains to have a full NMM domain.
:::
::::

The obtain figure is shown in the figure [below]

(theory-figure-many-mm)=
:::{figure} Figure-many-mm-3d.png

Ultimate domain NMM obtained as a set of slices computed as MM interaction domains.
:::
:::::

## Compute plane strain for given external forces

> With this algorithm, *structuralcodes*  computes the strain profile (represented by the triplet of values $\varepsilon_0$, $\chi_y$, $\chi_z$ that are respectivelly the axial strain corresponding to point $(0,0)$, and the curvatures respect axes $y$ and $z$) for a given set of external forces $N_{ext}$, $M_{y,ext}$, $M_{z,ext}$, with an iterative solution algorithm based on Newton-Rhapson method.

Reviewer: Fix typo and subscripts.

```suggestion
With this algorithm, *structuralcodes*  computes the strain profile (represented by the triplet of values $\varepsilon_0$, $\chi_y$, $\chi_z$ that are respectively the axial strain corresponding to point $(0,0)$, and the curvatures with respect to axes $y$ and $z$) for a given set of external forces $N_{\mathrm{ext}}$, $M_{\mathrm{y,ext}}$, $M_{\mathrm{z,ext}}$, with an iterative solution algorithm based on Newton-Rhapson method.
```

> The vector containing the internal force $N$, $M_y$, $M_z$ is obtained by integrating the stresses over the corr-section area of the structural element. These quantities can be expressed as follows:

Reviewer: Typo and subscripts.

```suggestion
The vector containing the internal force $N$, $M_{\mathrm{y}}$, $M_{\mathrm{z}}$ is obtained by integrating the stresses over the cross-section area. These quantities can be expressed as follows:
```

A similar consideration can be made for the other curvature $\chi_z$
:::

The stress $\sigma(y,z)$ can be related to the strain $\varepsilon(y,z)$ through the consitutive law (in general non-linear), according to:

:::{math}
:label: eq:stress-strain
\sigma(y,z) = f[\varepsilon(y,z)] = f[\varepsilon_0 + \chi_y \cdot z - \chi_z \cdot y]
:::

or, in vector form, as:

:::{math}
:label: eq:stress-strain_vector
\sigma(y,z)
=
f\left(
\begin{bmatrix} 1 & z & -y \end{bmatrix} \cdot 
\begin{bmatrix}
\varepsilon_0 \\
\chi_y \\
\chi_z
\end{bmatrix} 
\right)
:::

Substituting {eq}`eq:stress-strain` into {eq}`eq:section_forces`, the internal forces vecotr can be written as:

:::{math}
:label: eq:forces-strains
\begin{bmatrix}
N \\
M_y \\
M_z
\end{bmatrix}
=
\int_A
\begin{bmatrix}
1 \\
z \\
-y
\end{bmatrix}
f[\varepsilon_0 + \chi_y \cdot z - \chi_z \cdot y]
\, dA
:::

or, in vector form, as:

:::{math}
:label: eq:forces-strains-vector
\begin{bmatrix}
N \\
M_y \\
M_z
\end{bmatrix}
=
\int_A
\begin{bmatrix}
1 \\
z \\
-y
\end{bmatrix}
f\left(
\begin{bmatrix} 1 & z & -y \end{bmatrix} \cdot 
\begin{bmatrix}
\varepsilon_0 \\
\chi_y \\
\chi_z
\end{bmatrix} 
\right)
\, dA
:::

> In compact matrix notation {eq}`eq:forces-strains-vector`, calling the  can be written as:

Reviewer: Fix flow of sentence.

```suggestion
In compact matrix notation, {eq}`eq:forces-strains-vector` can be written as:
```

:::{math}
:label: eq:forces-strains-matrix
\mathbf{f}^{int}
=
\int_A
\mathbf{B}^T
\cdot
f\left(
\mathbf{B} \cdot \boldsymbol{\varepsilon}_g
\right)
:::

The vector $\boldsymbol{\varepsilon}_g$ represent the unkwon strain profile that we want to find for the given external forces $\mathbf{f}^{ext}$. To do so, a numeric method like Newton-Raphson can be used.

To this aim, the residual vector can be written as:

:::{math}
:label: eq:residuals
\begin{aligned}
\mathbf{R}(\boldsymbol{\varepsilon}_g) &= \mathbf{f}^{ext} - \mathbf{f}^{int} \\
  &= \mathbf{f}^{ext} - \int_A
\mathbf{B}^T
\cdot
f\left(
\mathbf{B} \cdot \boldsymbol{\varepsilon}_g
\right)
\end{aligned} 
:::

and linearizing it:

:::{math}
:label: eq:residuals_linearized
\mathbf{R}(\boldsymbol{\varepsilon}_g^{k+1}) = \mathbf{R}(\boldsymbol{\varepsilon}_g^k) + \mathbf{J}(\boldsymbol{\varepsilon}_g^k) \Delta \boldsymbol{\varepsilon}_g
:::

It is possible then to solve the current step:

:::{math}
:label: eq:nr_solve_current_step
\Delta \boldsymbol{\varepsilon}_g
= - \mathbf{J}^{-1}(\boldsymbol{\varepsilon}_g^k) 
\cdot
\mathbf{R}(\boldsymbol{\varepsilon}_g^k)
:::

> and therefore obtaning the new iteration of $\boldsymbol{\varepsilon}_g$:

Reviewer: Fix typo.

```suggestion
and therefore obtaining the new iteration of $\boldsymbol{\varepsilon}_g$:
```

# Section calculator

## General concept
Section calculators evaluate the structural response based on section properties and applied loads. They integrate results from the geometry and material models to provide insights into section performance by using a specific *section integrator*.

> The section calculator contains the methods that permits to handle a structual analysis for a cross section. For instance it is possible to compute the bending strength given the axial force, or to compute the moment curvatura, or to compute the strength domain. The outputs of those methods are represented as *ad-hoc* defined results objects which contains the data and the methods to further process them.

Reviewer: Minor typo: curvatura -> curvature.

And the result objects still do not contain much methods for further processing 😄

:::{Attention}
Note that in the current implementation both material constitutive laws and section are *unitless*. It means that user should choose a consistent set of units, like N for forces and mm for lengths; in this way the stresses will be in MPa and the moments in Nmm.
Pay attention that default-defined constitutive laws by material classes work with the units defined by the considered code, e.g. MPa for EC2 or MC2010.
:::

## Compute bending strength
With this algorithm, *structuralcodes* computes the bending strength of the section given the axial load (positive in tension and negative in compression) and an angle of the neutral axes respect to the y axis. 

![Bending Strength Rotated](FigureBendingRotated.png)

> In the rotated reference system **y\*z\***, the bending strength in terms of positive {math}`M_{\textrm{y}^*}` is computed.

Author: I am not sure I am following your suggestion. In the rotated reference system y*z*, Mz* is inherently zero because it is uniaxial bending in a rotated reference system...

    ![dN vs chi](FiguredN_chi.png)

    Bisection algorithm permits to find the zero of the function within some iterations.

    ![Bisection demo](bisection_demo.gif)

4. **Final Computation of Bending Strength**: once the equilibrium strain profile is found, the bending strength is calculated by integrating the obtained strain profile. This results in the final bending strength in the rotated coordinate system, which is then transformed back to the original coordinate system.

## Compute Moment-curvature relation

With this algorithm, *structuralcodes* computes the moment-curvature of the section given the axial load (positive in tension and negative in compression) and an angle of the neutral axes respect to the y axis. 

![Bending Strength Rotated](FigureBendingRotated.png)

The algorithm works with the following steps:

> 1. **If not provided, compute array of curvatures**: if curvatures are not provided, the array of curvatures corresponding to which computing the moments is computed. First the yield and 

Author: exactly! :)
